README.md
# WP PSR-3 Logger

This is a Wordpress plugin. It adds a standard PSR-3 logger. It's very handy whilst you're developing your plugins.

## How To Install

You need to install this as a 'must-use' plugin:

	cd wp-content/plugins
	git clone https://ganbarodigital/wp_psr3_logger
    cd ..
	mkdir mu-plugins
	ln -s plugins/wp_psr3_logger/mu/wp_psr3_logger_mu.php mu-plugins/

CHANGELOG.md
# CHANGELOG

## 1.0.0

Released Wed 16th November 2016.

### New

* Initial release.

src/autoload.php
<?php

/*
 * Exit if called directly.
 * PHP version check and exit.
 */
if ( ! defined( 'WPINC' ) ) {
    die;
}

function gd_wpsl_autoloader($className)
{
    // has a previous autoloader already loaded this?
    if (class_exists($className) || interface_exists($className) || trait_exists($className)) {
        return;
    }

    // the folders we will search
    static $psr0 = [
        "Psr\\Log\\" => realpath(__DIR__ . '/../vendor/psr/log'),
        "Monolog\Monolog\\" => realpath(__DIR__ . '/../vendor/monolog/monolog/src'),
    ];
    static $psr4 = [
        "GanbaroDigital\\ServiceLogger\\" => realpath(__DIR__ . '/../vendor/ganbarodigital/php-mv-service-logger'),
    ];

    // do we support the requested prefix?
    foreach ($psr0 as $prefix => $basePath) {
        if (substr($className, 0, strlen($prefix)) === $prefix) {
            // success :)
            $path = $basePath . '/' . str_replace(["_", "\\"], ["/", "/"], $className) . ".php";
            if (file_exists($path)) {
                require_once($path);
                return;
            }
        }
    }

    // no luck so far ... how about with PSR-4?
    foreach ($psr0 as $prefix => $basePath) {
        if (substr($className, 0, strlen($prefix)) === $prefix) {
            // success :)
            $path = $basePath . '/' . str_replace("\\", "/", $className) . ".php";
            if (file_exists($path)) {
                require_once($path);
                return;
            }
        }
    }

    // if we get here, then no success
}


wp_psr3_logger.php
<?php

/**
 * Plugin Name:       PSR3 Logger
 * Plugin URI:        https://github.com/ganbarodigital/wp_psr3_logger
 * Description:       This plugin provides a PSR3-compliant logger for your plugin to use
 * Version:           1.0.0
 * Author:            Ganbaro Digital
 * Author URI:        https://ganbarodigital.com
 * License:           3-Clause BSD
 * License URI:       http://www.opensource.org/licenses/bsd-license.php
 * Domain Path:       /languages
 * Text Domain:       wp_logging_service
 * Network:           true
 * GitHub Plugin URI: https://github.com/ganbarodigital/wp_psr3_logger
 * GitHub Branch:     develop
 * Requires WP:       4.0
 * Requires PHP:      5.6
 */

/*
 * Exit if called directly.
 * PHP version check and exit.
 */
if ( ! defined( 'WPINC' ) ) {
    die;
}

// this should make the code in our vendor/ folder available
require_once(__DIR__ . '/vendor/autoload.php');

// this creates our logger
require_once(__DIR__ . '/src/logging_service.php');


wp_logging_service.php
<?php
/**
 * Logging Service For PHP
 *
 * @package   WP Logging Service
 * @author    Stuart Herbert
 * @license   new BSD
 * @link      https://github.com/ganbarodigital/wp_logging_service
 */

/**
 * Plugin Name:       Wordpress Logging Service
 * Plugin URI:        https://github.com/ganbarodigital/wp_logging_service
 * Description:       This plugin provides a PSR3-compliant logger for your plugin to use
 * Version:           1.0.0
 * Author:            Ganbaro Digital
 * Author URI:        https://ganbarodigital.com
 * License:           3-Clause BSD
 * License URI:       http://www.opensource.org/licenses/bsd-license.php
 * Domain Path:       /languages
 * Text Domain:       wp_logging_service
 * Network:           true
 * GitHub Plugin URI: https://github.com/ganbarodigital/wp_logging_service
 * GitHub Branch:     develop
 * Requires WP:       4.0
 * Requires PHP:      5.6
 */

/*
 * Exit if called directly.
 * PHP version check and exit.
 */
if ( ! defined( 'WPINC' ) ) {
    die;
}

// this should make the code in our vendor/ folder available
require_once(__DIR__ . '/vendor/autoload.php');

// this creates our logger
require_once(__DIR_- . '/src/logging_service.php');


src/logging_service.php
<?php

use GanbaroDigital\ServiceLogger\V1\BuildServiceLogger;
use GanbaroDigital\ServiceLogger\V1\ServiceFormatter;
use GanbaroDigital\ServiceLogger\V1\ServiceLogger;
use Monolog\Handler\StreamHandler;
use Monolog\Processor\IntrospectionProcessor;
use Monolog\Processor\UidProcessor;

/*
 * Exit if called directly.
 * PHP version check and exit.
 */
if ( ! defined( 'WPINC' ) ) {
    die;
}

class PSR3_LOG
{
    /**
     * a PSR3-compliant logger for you to use
     *
     * @var ServiceLogger
     */
    public static $logger;

    /**
     * creates a new logger
     *
     * @return void
     */
    public static function init()
    {
        // here's the default config we're going to use, if none is
        // provided
        $defaultConfig = [
            'min_log_level' => 'INFO',
            'log_file' => ABSPATH . 'blog.log',
        ];

        // TODO: get overrides from the WP database
        //
        // we'll come back to this when we have an admin interface

        // build the component that will write log messages out
        $handler = new StreamHandler($defaultConfig['log_file'], $defaultConfig['min_log_level']);
        $handler->setFormatter(new ServiceFormatter);

        // build the logger itself
        self::$logger = new ServiceLogger(
            "ServiceLogger",
            [ $handler ],
            [ new IntrospectionProcessor ]
        );

        // all done
    }
}

PSR3_LOG::init();


mu/wp_logging_service_mu.php
<?php

/**
 * Copyright (c) 2016-present Ganbaro Digital Ltd
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the names of the copyright holders nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @category  Wordpress Plugin
 * @package   WP/LoggingService
 * @author    Stuart Herbert <stuherbert@ganbarodigital.com>
 * @copyright 2016-present Ganbaro Digital Ltd www.ganbarodigital.com
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @link      http://code.ganbarodigital.com/wp_logging_service
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
}

// make sure our logging plugin has been loaded
require_once(trailingslashit(WP_PLUGIN_DIR) . 'wp_logging_service/wp_logging_service.php');


mu/wp_psr3_logger_mu.php
<?php

/**
 * Copyright (c) 2016-present Ganbaro Digital Ltd
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the names of the copyright holders nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @category  Wordpress Plugin
 * @package   WP/LoggingService
 * @author    Stuart Herbert <stuherbert@ganbarodigital.com>
 * @copyright 2016-present Ganbaro Digital Ltd www.ganbarodigital.com
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @link      http://code.ganbarodigital.com/wp_logging_service
 */

/**
 * Plugin Name:       PSR3 Logger MU Loader
 * Plugin URI:        https://github.com/ganbarodigital/wp_psr3_logger
 * Description:       This plugin provides a PSR3-compliant logger for your plugin to use
 * Version:           1.0.0
 * Author:            Ganbaro Digital
 * Author URI:        https://ganbarodigital.com
 * License:           3-Clause BSD
 * License URI:       http://www.opensource.org/licenses/bsd-license.php
 * Domain Path:       /languages
 * Text Domain:       wp_psr3_logger
 * Network:           true
 * GitHub Plugin URI: https://github.com/ganbarodigital/wp_psr3_logger
 * GitHub Branch:     develop
 * Requires WP:       4.0
 * Requires PHP:      5.6
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
}

// make sure our logging plugin has been loaded
$wpLoggingServiceFile = 'wp_psr3_logger/wp_psr3_logger.php';
require_once(WP_PLUGIN_DIR . '/' . $wpLoggingServiceFile);

// avoid the plugin also being activated by the normal mechanism
add_action(
    'activated_plugin',
    function($plugin, $network_wide) use ($wpLoggingServiceFile) {
        if ($plugin === $wpLoggingServiceFile) {
            deactivate_plugins($wpLoggingServiceFile);
        }
    },
    10,
    2
);

// modify the plugins admin panel
function wpls_mu_plugin_active($actions)
{
    if (isset($actions['activate'])) {
        unset($actions['activate']);
    }
    if (isset($actions['delete'])) {
        unset($actions['delete']);
    }
    if (isset($actions['deactivate'])) {
        unset($actions['deactivate']);
    }

    return array_merge([
        'mu-plugin' => esc_html__('Activated as mu-plugin', 'wp_logging_service')
    ], $actions);
}


add_filter('network_admin_plugin_action_links_' . $wpLoggingServiceFile, 'wpls_mu_plugin_active' );
add_filter('plugin_action_links_' . $wpLoggingServiceFile, 'wpls_mu_plugin_active' );
add_action(
    'after_plugin_row_' . $wpLoggingServiceFile,
    function () {
        print('<script>jQuery(".inactive[data-plugin=\'' . $wpLoggingServiceFile . '\']").attr("class", "active");</script>' );
        print('<script>jQuery(".active[data-plugin=\'' . $wpLoggingServiceFile . '\'] .check-column input").remove();</script>' );
    }
);